doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,246 @@
+
+There is a corresponding MDIO access IPC client code in the form of dynamic link library which provides the flexiblity to load the library at runtime. Assuming the MDIO access library for sysfs is also in the form of dynamic library, gbsyncd can select the MDIO access library at runtime based on some configuration in the gearbox\_config.json file.
+
+The same gearbox\_config.json file already has the information of the PAI library name. The information can be used to dynamically load the PAI library at runtime.
+
+The informationn of PHY device using clause 22 mdio can also be added to the gearbox\_config.json file. It will help to decide the mdio access mode at runtime.
+
+![syncd gbsyncd MDIO IPC](images/MDIO-IPC.png)
+
+The VendorSai class constructor is overloaded to handle the dynamic linking of the libraries. The  syncd class VendorSai currently only has a constructor with no argument. The linked SAI library path is determined at syncd build time. New constructors with arguments for the VendorSai class is introduced as the overloading constructors. The new constructors can either have the SAI/PAI library path as the argument or have PHY id as one of the arguments. When the constructor with PHY id as the argument is used, the syncd can index into the "phys" section of gearbox configuration file "gearbox\_config.json" to get the PAI library path and MDIO access library path at run time.
+
+### High-Level Design 
+
+The high level design of the gbsyncd mdio access function using the mdio bus from the switch NPU and using single gbsyncd docker in SONiC image can be covered in the following points:
+		
+	- It is an enhancement to the sonic gearbox framework.
+	- Syncd, SAI and some sonic device files are modified.
+	- There are 3 repositories that would be changed, sonic-sairedis, SAI and sonic-buildimage.
+	- gbsyncd will have dependencies on syncd to create the Unix IPC socket. 
+	- Syncd will have new classes for the MDIO IPC server and for the parser of gearbox configuration.
+	- There is a new MDIO IPC client library is added.
+	- External PHY firmware download time always has performance requirements/impact.
+	- For debugging, beside the syslog logging, the Unix socket IPC mechanism can be simulated by socat.
+	- The current change is only limited to the Broadcom switch NPU platform.
+	- SAI switch api has added the mido clause 22 access functions.
+	- The SaiInterface class has added the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- The VendoSai class overrides the mdio clause 45 read/write and mdio clause 22 read/write virtual functions.
+	- New keys are added to the gearbox configuration file "gearbox\_config.json".
+
+When the VendorSai class constructor with PHY id argument is used in syncd.cpp, we can assume a simple case where syncd context number is the same as the PHY id.
```

Comment:
Now one syncd process associates with one context entry in context_config.json. Will it launch multiple syncd processes in new design?

Reply:
This design does not aggregate PHYs. It also does not block other software components/changes to manage multiple PHYs in a single context.

Reply:
> assume a simple case where syncd context number is the same as the PHY id

Still do not understand the meaning of the assumption. There is only one context for gbsyncd.

Reply:
The original gearbox design has one context for each PHY. That is why there is ongoing effort to aggregate multiple PHY in a single context. Our design is based on what is currently in the community sonic. There is different way to aggregate PHYs into a single context. It is outside the scope of this design document.